### PR TITLE
Pia 3027 allow to add custom loading indicator with a custom animation

### PR DESCRIPTION
**Pull request by vizaknaidavid**
For testing it, you can uncomment the custom loading indicator configuration from the BankSDK example appcoordinator.

Simulator Screen Shot - iPad Pro (12 9-inch) (5th generation) - 2022-09-12 at 15 00 06

### CODE REVIEW

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Analysis/AnalysisViewController.swift [hunk 1]
```diff
@@ -244,6 +249,65 @@ import UIKit
...
+        addLoadingView(intoContainer: loadingIndicatorContainer)
+
+        if let loadingIndicator = giniConfiguration.analysisScreenLoadingIndicator {
+            addLoadingText(below: loadingIndicator.injectedView())
+            loadingIndicator.startAnimation()
+        } else {
+            addLoadingText(below: loadingIndicatorView)
+            loadingIndicatorView.startAnimating()
+        }
+    }
```
Comment (chriskryn) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Analysis/AnalysisViewController.swift:265]: @vizaknaidavid @zladzeyka  Could you tell me how custom loading indicator should work? Should we allow displaying both at the same time ? Or one replaces another?

**Review by chriskryn — CHANGES_REQUESTED** [review]
There are some linter issues in AnalysisViewController

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Analysis/AnalysisViewController.swift (hunk 1 above)
Comment (zladzeyka) [reply CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Analysis/AnalysisViewController.swift:265]: @chriskryn  there should be 1 loading indicator 😎: default or with a custom view.

Comment (zladzeyka) [reply CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Analysis/AnalysisViewController.swift:265]: @vizaknaidavid there is no need to add a custom view for a loading indicator to a super view if giniConfiguration.analysisScreenLoadingIndicator is not set,

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Analysis/AnalysisViewController.swift [hunk 2]
```diff
@@ -244,6 +249,65 @@ import UIKit
...
+        if let customLoadingIndicator = giniConfiguration.analysisScreenLoadingIndicator?.injectedView() {
+            loadingIndicator = customLoadingIndicator
+        } else {
+            loadingIndicator = loadingIndicatorView
+        }
+
+        if let container = container {
+            container.translatesAutoresizingMaskIntoConstraints = false
+            loadingIndicator.translatesAutoresizingMaskIntoConstraints = false
+            self.view.addSubview(container)
```
Comment (chriskryn) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Analysis/AnalysisViewController.swift:289]: @vizaknaidavid  Could you remove the 'self' when it's not necessary?

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Analysis/AnalysisViewController.swift (hunk 1 above)
Comment (chriskryn) [reply CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Analysis/AnalysisViewController.swift:265]: @zladzeyka  Ok  :slightly_smiling_face: , I think it is used to center both indicator and text within the screen.

Comment (chriskryn) [reply CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Analysis/AnalysisViewController.swift:265]: @vizaknaidavid @zladzeyka So do we need a text under custom indicator ? Then I would think that the custom indicator should be enough .

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Analysis/AnalysisScreenLoadingIndicatorAdapter.swift [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+//  Created by David Vizaknai on 12.09.2022.
+//
+
+import Foundation
+/**
+*   Adapter for injecting a custom loading indicator for the analysis viewcontroller.
+*/
+public protocol AnalysisScreenLoadingIndicatorAdapter: InjectedViewAdapter {
+    /**
+     *  Called when the screen is loaded, can be also called from the AnalysisViewController's showAnimation() method.
```
Comment (a-szotyori) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Analysis/AnalysisScreenLoadingIndicatorAdapter.swift:14]: For me the intention of the documentation here is to inform clients about what should they do when implementing this method. The questions the documentation should answer is: "As a client what do I have to do in this method? When is this method called? Why do I have to implement it?"

We don't need to mention the `showAnimation()` method here. It is currently public, but we'll have to make it internal when we remove the Component API. Clients won't be able to interact directly with our screen view controllers anymore.

My change suggestion would be:
```suggestion
     *  Called when the screen is loaded. You should start the loading indicator animation in this method.
```

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Analysis/AnalysisScreenLoadingIndicatorAdapter.swift [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+     */
+    func startAnimation()
+    /**
+     *  Can be called from the AnalysisViewController's hideAnimation() method.
```
Comment (a-szotyori) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Analysis/AnalysisScreenLoadingIndicatorAdapter.swift:18]: Similarly to the above comment my change proposal would be:
```suggestion
     *  Called when the screen has disappeared. You should stop the loading indicator animation in this method.
```

**Review by a-szotyori — CHANGES_REQUESTED** [review]
Looks great. I only have comments regarding the documentation.

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Analysis/AnalysisViewController.swift (hunk 1 above)
Comment (vizaknaidavid) [reply CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Analysis/AnalysisViewController.swift:265]: @zladzeyka I add a container that is centered in the screen in which I add the loading indicator and the text, so they are centered in the screen. If the custom indicator is not set, the conatiner will contain the default loading and the text.

Comment (vizaknaidavid) [reply CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Analysis/AnalysisViewController.swift]: So can this thread be resolved without changes? @chriskryn  @zladzeyka  :D

Comment (zladzeyka) [reply CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Analysis/AnalysisViewController.swift]: Yes, I think so ;)

**Review by chriskryn — APPROVED** [review]
Looks Good! :+1:

**Review by a-szotyori — APPROVED** [review]

**Pull request merged**